SCAutolib/models/CA.py

```diff
@@ -196,3 +207,322 @@ def cleanup(self):
+            logger.error(result.stderr)
+            raise SCAutolibException(msg)
+        logger.debug("File for setting up IPA client for smart cards is "
+                     f"copied to {ipa_client_script}")
+        return ipa_client_script
+
+    def request_cert(self, csr: Path, username: str, cert_out: Path):
+        """
+        Request certificate using CSR from IPA CA for given username. It is
+        a wrapper around the python_freeipa.client_meta.ClientMeta.cert_request
+        method. It works with a file, extracts CSR data from it and then
+        stores in PEM format adding required prefix and suffix as in normal
+        certificate and. If cert_out is a directory, then certificate would be
+        stored in this directory with name <username>.pem. If it is a file,
+        then check if it has PEM extension. If not, append .pem suffix to the
+        name.
+
+        :param csr: path to CSR
+        :type csr: patlib.Path
+        :param username: subject for the certificate
+        :type username: str
+        :param cert_out: path where the certificate is stored. Can be a
+                         directory or a file.
+        :type cert_out: patlib.Path
+
+        :return: Path to the PEM certificate.
+        :rtype: patlib.Path
+        """
+        with csr.open() as f:
+            csr_content = f.read()
+        r = self.meta_client.cert_request(a_csr=csr_content,
+                                          o_principal=username)
+        logger.debug(r)
+        cert = r["result"]["certificate"]
+
+        if cert_out.is_dir():
+            cert_out = cert_out.joinpath(f"{username}.pem")
+        else:
+            cert_out = cert_out.with_suffix(".pem")
+
+        with cert_out.open("w") as f:
+            f.write("-----BEGIN CERTIFICATE-----\n"
+                    f"{cert}\n"
+                    f"-----END CERTIFICATE-----")
+        return cert_out
+
+    def add_user(self, user):
+        """
+        Add given user to IPA server. It is a wrapper on the
+        python_freeipa.client_meta.ClientMeta.user_add method. Just extracts
+        necessary fields from IPAUser objet and pass them to the method. As a
+        result, o_givenname == o_uid == o_sn == o_cn for simplicity.
+
+        :param user: User to be added to the IPA server.
+        """
+        r = self.meta_client.user_add(user.username, user.username,
+                                      user.username, user.username,
+                                      o_userpassword=user.password)
+        logger.debug(r)
+        logger.info(f"User {user.username} is added to the IPA server")
+
+    def del_user(self, user):
+        """
+        Remove user from IPA server.
+        :param user: User to be deleted
+
+        :raise AssertionError: If operation is failed. Fail is detected in
+            return value from request to IPA server.
+        """
+        r = self.meta_client.user_del(user.username)["result"]
```

Comment:
Might need error handling, not sure if user_del call can fail

Reply:
it is already done by python_freeipa module